File: lab1/lab1_5.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

void QuickSort(vector<int>& A, int left, int right) {
    int i = left;
    int j = right;
    int pivot = A[(left + right) / 2];

    while (i <= j) {
        while (A[i] < pivot) {
            i++;
        }

        while (A[j] > pivot) {
            j--;
        }

        if (i <= j) {
            swap(A[i], A[j]);
            i++;
            j--;
        }
    }

    if (left < j) {
        QuickSort(A, left, j);
    }

    if (i < right) {
        QuickSort(A, i, right);
    }
}

int main() {
    int N;
    cin >> N;

    vector<int> A(N);

    for (int i = 0; i < N; i++) {
        cin >> A[i];
    }

    if (N > 0) {
        QuickSort(A, 0, N - 1);
    }

    for (int i = 0; i < N; i++) {
        cout << A[i] << " ";
    }

    return 0;
}

File: lab1/lab1_5.py
import sys


def QuickSort(A, left, right):
    i = left
    j = right
    pivot = A[(left + right) // 2]

    while i <= j:
        while A[i] < pivot:
            i += 1

        while A[j] > pivot:
            j -= 1

        if i <= j:
            A[i], A[j] = A[j], A[i]
            i += 1
            j -= 1

    if left < j:
        QuickSort(A, left, j)

    if i < right:
        QuickSort(A, i, right)


data = list(map(int, sys.stdin.read().split()))
n = data[0]
A = data[1:1 + n]

if n > 0:
    QuickSort(A, 0, n - 1)

print(*A)
